### ISSUE TITLE
needs should not alter regular install approach: one after the other

### ISSUE THREAD

**Issue by Morriz** [helmfile/helmfile#208]
### Operating system

Ubuntu 20.004

### Helmfile Version

0.145.0

### Helm Version

version.BuildInfo{Version:"v3.8.1", GitCommit:"5cb9af4b1b271d11d7a97a71df3ac337dd94ad37", GitTreeState:"clean", GoVersion:"go1.17.8"}

### Bug description

When providing `needs:` we see that releases don't get installed one after the other when using `apply`, but instead get errors about crds not being available (which are to be applied by chart A, and consume by chart B).

Is this due to diff running on ALL the deps beforehand? (DAG resolution happening beforehand?) I think that should not happen, and every release should just be diffed and installed first, before diffing the next.

### Example helmfile.yaml

releases:
releases:
  - name: gatekeeper
    installed: true
    namespace: gatekeeper-system
    chart: ../charts/gatekeeper
    disableValidationOnInstall: true
    labels:
      pkg: gatekeeper
    values: ...
  - name: gatekeeper-artifacts
    installed: true
    needs: [gatekeeper]
    namespace: gatekeeper-system
    chart: ../charts/gatekeeper-artifacts
    labels:
      pkg: gatekeeper
    values: ...

### Error message you've seen (if any)

basically errors about crds not being available when rendering the next in the DAG.

### Steps to reproduce

https://github.com/redkubes/otomi-core

**Comment by Morriz** [comment]
I have not tested `sync` which bypasses `diff`. Will try now

**Comment by Morriz** [comment]
it works fine with diff, so my observation might be correct

**Comment by mumoshu** [comment]
@Morriz Hey!

> basically errors about crds not being available when rendering the next in the DAG.

`disableValidationInstall` has been implemented exactly for that. 

> s this due to diff running on ALL the deps beforehand? (DAG resolution happening beforehand?) 

FWIW, this is working as intended. Perhaps you remember helmfile has `--interactive` flag that instructs helmfile-apply to prompt for changes after showing diff but before sync? It enables a workflow very similar to that of `terraform apply` without `--auto-approve`. If it had to show diff before each node(a subset of all the releases) in a dependency graph, it needs to prompt per node, multiple times in a single `helmfile apply` run. That's not what we wanted.

I think this is almost a limitation stemming from helm-diff that there's no way to provide CRDs that are not present in the cluster yet but will be created BEFORE the diff target is installed onto the cluster. If helm-diff had a flag like `--additional-crds`, we might be able to call it like, for example, `helm diff upgrade --additioinal-crds charts/promehetus-operator/crds/servicemonitors.aml myapp charts/myapp` so that our app that depends on the ServiceMonitor CRD won't result in diff error even though the prometheus-operator that provides the CRD isn't installed yet. As long as helm-diff provided that, it should be very easy for helmfile to compute and pass correct `additioinal-crds` to helm-diff according to the `needs`.

**Comment by Morriz** [comment]
I get it now. That surely would be very handy indeed. Unfortunately we will still need `disableValidationInstall` even afterwards, as some charts create crds from template crds (as is the case with gatekeeper), which helm doesn't know about.
I will close this for now.

**Comment by Morriz** [comment]
Sure. We are not the makers of Gatekeeper btw. But those CRDs are crafted by their solution on the fly from crd templates icw constraints. It takes a couple of seconds for each to become ready, so we added a post-install hook that waits for those to become ready so the last chart can actually create the final CRs.

**Comment by mumoshu** [comment]
@Morriz Hey. I took a glance on how the gatekeeper works and unable to locate the dynamic CRD templates you mentioned. Instead, I found https://github.com/open-policy-agent/gatekeeper/blob/master/charts/gatekeeper/templates/upgrade-crds-hook.yaml. Is this what we are talking about?

If so, apparently, it's not dynamically generating CRDs. It looks like just a handy feature to upgrade CRDs via helm-upgrade, which isn't possible with the vanilla helm-upgrade. You may already know but helm-upgrade never upgrade CRDs according to what contained in the `crds/` directory of the new version of the chart after the first install.
